### ISSUE TITLE
Type of `CryptoKey.algorithm` is incomplete, making usage of `SubtleCrypto` difficult

### ISSUE THREAD

**Issue by LiraNuna** [Microsoft/TypeScript#21609]
**Code**

```ts
function sign(key: CryptoKey, message: ArrayBuffer): Promise<ArrayBuffer> {
    return crypto.subtle.sign(
        key.algorithm,
        key,
        message,
    );
}
```

**Expected behavior:**
`CryptoKey`.algorithm is an appropriate way to pass the key algorithm to subtle methods such as `sign` and `encrypt`. However, the original definition of the type `CryptoKey` defines the type of `algorithm` as `KeyAlgorithm` which misses extra attributes that could be there.

Therefor the expected behavior is for the code to now have a type error and compile correctly. Maybe the correct type for `CryptoKey` should be 

```ts
interface CryptoKey {
    readonly algorithm: RsaKeyAlgorithm | EcKeyAlgorithm | AesKeyAlgorithm | HmacKeyAlgorithm | DhKeyAlgorithm;
    readonly extractable: boolean;
    readonly type: string;
    readonly usages: string[];
}
```

I am not sure how to fully test such a change, since modifying `lib.es6.d.ts` locally and running `tsc` produces the same message.

**Actual behavior:**
The following type error:
```
Argument of type 'KeyAlgorithm' is not assignable to parameter of type 'string | RsaPssParams | EcdsaParams | AesCmacParams'.
  Type 'KeyAlgorithm' is not assignable to type 'AesCmacParams'.
  Property 'length' is missing in type 'KeyAlgorithm'. (2345)
```

**Playground Link:** [Playground link](https://www.typescriptlang.org/play/#src=function%20sign(key%3A%20CryptoKey%2C%20message%3A%20ArrayBuffer)%3A%20Promise%3CArrayBuffer%3E%20%7B%0D%0A%20%20%20%20return%20crypto.subtle.sign(%0D%0A%20%20%20%20%20%20%20%20key.algorithm%2C%0D%0A%20%20%20%20%20%20%20%20key%2C%0D%0A%20%20%20%20%20%20%20%20message%2C%0D%0A%20%20%20%20)%3B%0D%0A%7D)

**Related Issues:**

* Unknown

**Comment by mhegazy** [comment]
PRs welcomed. You can find more information about contributing lib.d.ts fixes at https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md#contributing-libdts-fixes.

**Comment by LiraNuna** [comment]
I tried giving this a go but it required me to go down into a rabbit hole I was not able to fully grok. 
My proposed changes did not fix the issue and therefor I'm not sure what a "correct" solution may look like.

**Comment by LiraNuna** [comment]
This issue has been open for 4 months without much progress. I've been trying to address this issue myself but I do not have the knowledge necessary to contribute to `lib.d.ts` even with the instructions (especially since this depends on Edge?).

Is there any plan to address this? This makes WebCrypto quite unusable on TypeScript.

**Comment by jiripudil** [comment]
Hello, I'm currently enhancing the Web Crypto API types and I have mixed feelings about this issue. While I agree that it might be useful to have a more specific set of types for `CryptoKey.algorithm`, I don't understand the motivation.

> `CryptoKey.algorithm` is an appropriate way to pass the key algorithm to subtle methods such as `sign` and `encrypt`

This is what confuses me the most because I can't agree with it. More often than not, `SubtleCrypto` methods require a different set of properties from those contained in `KeyAlgorithm`: commonly additional details such as initialization vectors are required, while some information in `KeyAlgorithm` such as key length is excessive because that's just information about the key, not parameters of the crypto operation.

**Comment by LiraNuna** [comment]
> This is what confuses me the most because I can't agree with it. More often than not, SubtleCrypto methods require a different set of properties from those contained in KeyAlgorithm

Sure, I agree with that. For that, wouldn't you want to use `{...key.algorithm, more: rest}` which will have similar issues?

**Comment by jakebailey** [comment]
This was fixed in #37464.
